bundles/org.openhab.binding.livisismarthome/src/main/java/org/openhab/binding/livisismarthome/internal/client/LivisiClient.java

```diff
@@ -0,0 +1,479 @@
+import org.openhab.binding.livisismarthome.internal.handler.LivisiBridgeConfiguration;
+import org.openhab.core.auth.client.oauth2.AccessTokenResponse;
+import org.openhab.core.auth.client.oauth2.OAuthClientService;
+import org.openhab.core.auth.client.oauth2.OAuthException;
+import org.openhab.core.auth.client.oauth2.OAuthResponseException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonSyntaxException;
+
+/**
+ * The main client that handles the communication with the LIVISI SmartHome API service.
+ *
+ * @author Oliver Kuhl - Initial contribution
+ * @author Hilbrand Bouwkamp - Refactored to use openHAB http and oauth2 libraries
+ * @author Sven Strohschein - Renamed from Innogy to Livisi
+ */
+@NonNullByDefault
+public class LivisiClient {
+
+    private static final String BASIC = "Basic ";
+    private final Logger logger = LoggerFactory.getLogger(LivisiClient.class);
+
+    /**
+     * date format as used in json in API. Example: 2016-07-11T10:55:52.3863424Z
+     */
+    private static final String DATE_FORMAT = "yyyy-MM-dd'T'HH:mm:ss.SSSZ";
+
+    private final Gson gson = new GsonBuilder().setDateFormat(DATE_FORMAT).create();
+    private final LivisiBridgeConfiguration bridgeConfiguration;
+    private final OAuthClientService oAuthService;
+    private final URLConnectionFactory connectionFactory;
+
+    public LivisiClient(final LivisiBridgeConfiguration bridgeConfiguration, final OAuthClientService oAuthService,
+            final URLConnectionFactory connectionFactory) {
+        this.bridgeConfiguration = bridgeConfiguration;
+        this.oAuthService = oAuthService;
+        this.connectionFactory = connectionFactory;
+    }
+
+    /**
+     * Gets the status
+     * As the API returns the details of the SmartHome controller (SHC), the config version is returned.
+     *
+     * @return config version
+     */
+    public String refreshStatus() throws IOException, ApiException, AuthenticationException {
+        logger.debug("Get LIVISI SmartHome status...");
+        final StatusResponseDTO status = executeGet(URLCreator.createStatusURL(bridgeConfiguration.host),
+                StatusResponseDTO.class);
+
+        final String configVersion = status.getConfigVersion();
+        logger.debug("LIVISI SmartHome status loaded. Configuration version is {}.", configVersion);
+        return configVersion;
+    }
+
+    /**
+     * Executes a HTTP GET request with default headers and returns data as object of type T.
+     *
+     * @param url request URL
+     * @param clazz type of data to return
+     * @return response content
+     */
+    private <T> T executeGet(final String url, final Class<T> clazz)
+            throws IOException, AuthenticationException, ApiException {
+
+        HttpURLConnection connection = createBaseRequest(url, HttpMethod.GET);
```

Comment:
I tried to use the jetty http client (by using HttpClientFactory#getCommonHttpClient()), but this results in EOFException. The Jetty http client seems to have problems with concurrent requests, see also other issue reports where developers ran into the same situation, for example https://github.com/eclipse/jetty.project/issues/3891 and https://github.com/eclipse/jetty.project/issues/3485. Therefore I decided to just use the simple JDK build-in functionalities... I could try to create an own http client instance (instead of using the shared http client), but would that be better?

Reply:
Did you try to pass along the object you already retrieved in `LivisiHandlerFactory`:

```
    @Activate
    public LivisiHandlerFactory(@Reference OAuthFactory oAuthFactory, @Reference HttpClientFactory httpClientFactory) {
        this.oAuthFactory = oAuthFactory;
        httpClient = httpClientFactory.getCommonHttpClient();
    }
```

?

For example it could be added as a constructor argument for `LivisiClient` (replacing `URLConnectionFactory`) and added here in `LivisiHandlerFactory`:

```
    LivisiClient createClient(final OAuthClientService oAuthService) {
        return new LivisiClient(bridgeConfiguration, oAuthService, new URLConnectionFactory());
    }
```

I.e.:

```
    LivisiClient createClient(final OAuthClientService oAuthService) {
        return new LivisiClient(bridgeConfiguration, oAuthService, httpClient);
    }
```

This issues you are referencing were resolved years ago.

Reply:
Yes, I tried it that way. I could push this change to a separated branch if that is interesting.

Reply:
Sure, I might be able to spot something from it (or not). :-) If you still have it, I can have a look.

Reply:
Here is the branch (based on OpenHAB 3.2) where you can see the change: https://github.com/Novanic/openhab-addons/tree/livisi-httpclient

Here is the log / Exception. It seems to be a concurrency problem. When I slow down the request execution, everything works fine...

`22:37:46.199 [DEBUG] [marthome.internal.client.LivisiClient] - LIVISI SmartHome status loaded. Configuration version is 680.
22:37:46.200 [DEBUG] [marthome.internal.client.LivisiClient] - Loading locations...
22:37:46.208 [DEBUG] [.internal.handler.LivisiBridgeHandler] - IOException occurred
java.io.IOException: java.util.concurrent.ExecutionException: java.io.EOFException: HttpConnectionOverHTTP@190551f5::SocketChannelEndPoint@4dfe810a{l=/192.168.178.
58:53728,r=/192.168.178.41:8080,ISHUT,fill=-,flush=-,to=3/0}{io=0/0,kio=0,kro=1}->HttpConnectionOverHTTP@190551f5(l:/192.168.178.58:53728 <-> r:/192.168.178.41:808
0,closed=false)=>HttpChannelOverHTTP@d94436(exchange=HttpExchange@5ce082e8{req=HttpRequest[GET /location HTTP/1.1]@6f59ea78[TERMINATED/null] res=HttpResponse[null 
0 null]@5696d048[PENDING/null]})[send=HttpSenderOverHTTP@1c24c658(req=QUEUED,snd=COMPLETED,failure=null)[HttpGenerator@440be0f6{s=START}],recv=HttpReceiverOverHTTP
@6421c825(rsp=IDLE,failure=null)[HttpParser{s=CLOSED,0 of -1}]]
        at org.openhab.binding.livisismarthome.internal.client.LivisiClient.executeRequest(LivisiClient.java:175) ~[?:?]
        at org.openhab.binding.livisismarthome.internal.client.LivisiClient.executeGet(LivisiClient.java:127) ~[?:?]
        at org.openhab.binding.livisismarthome.internal.client.LivisiClient.executeGetList(LivisiClient.java:140) ~[?:?]
        at org.openhab.binding.livisismarthome.internal.client.LivisiClient.getLocations(LivisiClient.java:373) ~[?:?]
        at org.openhab.binding.livisismarthome.internal.manager.FullDeviceManager.createLocationMap(FullDeviceManager.java:125) ~[?:?]
        at org.openhab.binding.livisismarthome.internal.manager.FullDeviceManager.getFullDevices(FullDeviceManager.java:63) ~[?:?]
        at org.openhab.binding.livisismarthome.internal.manager.DeviceStructureManager.refreshDevices(DeviceStructureManager.java:80) ~[?:?]
        at org.openhab.binding.livisismarthome.internal.handler.LivisiBridgeHandler.refreshDevices(LivisiBridgeHandler.java:209) ~[?:?]
        at org.openhab.binding.livisismarthome.internal.handler.LivisiBridgeHandler.startClient(LivisiBridgeHandler.java:183) ~[?:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]
Caused by: java.util.concurrent.ExecutionException: java.io.EOFException: HttpConnectionOverHTTP@190551f5::SocketChannelEndPoint@4dfe810a{l=/192.168.178.58:53728,r
=/192.168.178.41:8080,ISHUT,fill=-,flush=-,to=3/0}{io=0/0,kio=0,kro=1}->HttpConnectionOverHTTP@190551f5(l:/192.168.178.58:53728 <-> r:/192.168.178.41:8080,closed=f
alse)=>HttpChannelOverHTTP@d94436(exchange=HttpExchange@5ce082e8{req=HttpRequest[GET /location HTTP/1.1]@6f59ea78[TERMINATED/null] res=HttpResponse[null 0 null]@56
96d048[PENDING/null]})[send=HttpSenderOverHTTP@1c24c658(req=QUEUED,snd=COMPLETED,failure=null)[HttpGenerator@440be0f6{s=START}],recv=HttpReceiverOverHTTP@6421c825(
rsp=IDLE,failure=null)[HttpParser{s=CLOSED,0 of -1}]]
        at org.eclipse.jetty.client.util.FutureResponseListener.getResult(FutureResponseListener.java:118) ~[?:?]
        at org.eclipse.jetty.client.util.FutureResponseListener.get(FutureResponseListener.java:101) ~[?:?]
        at org.eclipse.jetty.client.HttpRequest.send(HttpRequest.java:730) ~[?:?]
        at org.openhab.binding.livisismarthome.internal.client.LivisiClient.executeRequest(LivisiClient.java:167) ~[?:?]
        ... 14 more
Caused by: java.io.EOFException: HttpConnectionOverHTTP@190551f5::SocketChannelEndPoint@4dfe810a{l=/192.168.178.58:53728,r=/192.168.178.41:8080,ISHUT,fill=-,flush=
-,to=3/0}{io=0/0,kio=0,kro=1}->HttpConnectionOverHTTP@190551f5(l:/192.168.178.58:53728 <-> r:/192.168.178.41:8080,closed=false)=>HttpChannelOverHTTP@d94436(exchang
e=HttpExchange@5ce082e8{req=HttpRequest[GET /location HTTP/1.1]@6f59ea78[TERMINATED/null] res=HttpResponse[null 0 null]@5696d048[PENDING/null]})[send=HttpSenderOve
rHTTP@1c24c658(req=QUEUED,snd=COMPLETED,failure=null)[HttpGenerator@440be0f6{s=START}],recv=HttpReceiverOverHTTP@6421c825(rsp=IDLE,failure=null)[HttpParser{s=CLOSE
D,0 of -1}]]
        at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.earlyEOF(HttpReceiverOverHTTP.java:376) ~[?:?]
        at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:1614) ~[?:?]
        at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.shutdown(HttpReceiverOverHTTP.java:259) ~[?:?]
        at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.process(HttpReceiverOverHTTP.java:181) ~[?:?]
        at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.receive(HttpReceiverOverHTTP.java:79) ~[?:?]
        at org.eclipse.jetty.client.http.HttpChannelOverHTTP.receive(HttpChannelOverHTTP.java:131) ~[?:?]
        at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.onFillable(HttpConnectionOverHTTP.java:172) ~[?:?]
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311) ~[?:?]
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105) ~[?:?]
        at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104) ~[?:?]
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:338) ~[?:?]
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:315) ~[?:?]
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:173) ~[?:?]
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131) ~[?:?]
        at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:386) ~[?:?]
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883) ~[?:?]
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034) ~[?:?]
        ... 1 more`

Reply:
LivisiClient is the one and only class which sends requests. But OpenHAB sends of course some other requests, for example via OAuthClientService which is used within LivisiBridgeHandler.

The Jetty-HttpClient seems to have start and stop methods, which seem to mean that it holds a state (see https://zetcode.com/java/jetty/httpclient/ )... I don't know why a HttpClient has to hold a state... And that could mean that there could be still concurrency issues, like other users are reporting too.

Are there other HttpClients which are allowed? For example Spring RestTemplate (which don't require start and stop methods or something like that). I would think to use build-in JDK methods, like I did, would also not be a bad practice.

Reply:
I have now tried it out again. synchronized for executeRequest doesn't help.

1. The oauth requests are successful
2. The /status request is successful
3. The /location request is failing

It is also difficult to debug, because the Request#send() method is already crashing. Therefore I get no result and would have to debug internal logics of Jetty-HttpClient. When I set a breakpoint between the /status and /location requests, the /location request is successful too. Therefore it seems to be an issue with the Jetty-HttpClient or with the backend (but when it is the backend, why do the build-in JDK methods and other clients don't have issues with it?).

Reply:
Let's leave it like it is for now, since it's working and well tested.